File: src/fragment/SpellPart.ts
import { Text } from 'pixi.js';
import Fragment, { FragmentType, register } from './Fragment';
import PatternGlyph from './PatternGlyph';
import {
    StructEndecBuilder,
    PrimitiveEndecs,
    StructEndec,
    Optional,
    SerializationContext,
    BufferDeserializer,
    listOf,
} from 'KEndec';
import {
    PROTOCOL_VERSION_ATTRIBUTE,
    protocolVersionAlternatives,
    recursive,
    UBER_COMPACT_ATTRIBUTE,
    withAlternative,
} from '~/endecTomfoolery';
import { SpellInstruction } from '~/spellInstruction';
import { SpellUtils } from '~/SpellUtils';

type Buffer = any;

export default class SpellPart extends Fragment {
    glyph: Fragment;
    subParts: ReadonlyArray<SpellPart>;

    constructor(glyph?: Fragment, subParts?: ReadonlyArray<SpellPart>) {
        super();

        this.glyph = glyph != undefined ? glyph : new PatternGlyph();
        this.subParts = subParts != undefined ? subParts : [];
    }

    override toString(): string {
        return 'spell';
    }

    override type(): FragmentType<SpellPart> {
        return SPELL_PART;
    }

    getSubParts(): ReadonlyArray<SpellPart> {
        return this.subParts;
    }

    static fromBytesOld(protocolVersion: number, buf: Buffer): SpellPart {
        return SPELL_PART.endec.decode(
            SerializationContext.empty().withAttributes([
                UBER_COMPACT_ATTRIBUTE,
                PROTOCOL_VERSION_ATTRIBUTE.instance(protocolVersion),
            ]),
            new BufferDeserializer(buf)
        );
    }
}

const SPELL_PART = register(
    'spell_part',
    0xaa44aa,
    recursive((self: StructEndec<SpellPart>) =>
        StructEndecBuilder.of2(
            Fragment.ENDEC.fieldOf('glyph', (fragment: SpellPart) => fragment.glyph),
            protocolVersionAlternatives(
                new Map([[1, self.listOf()]]),
                withAlternative(
                    SpellInstruction.stackEndec().xmap(
                        (instructions) =>
                            SpellUtils.decodeInstructions(instructions, [], [], Optional.empty()),
                        SpellUtils.flattenNode
                    ),
                    self
                ).listOf()
            ).fieldOf('sub_parts', (fragment: SpellPart) =>
                listOf([...fragment.subParts])
            ),
            (glyph, subparts) => new SpellPart(glyph, subparts.asJsReadonlyArrayView())
        )
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/SpellPart.kt
package me.maplesyrum.tricksterstudio.spell.fragment


import me.maplesyrum.tricksterstudio.SpellUtils
import me.maplesyrum.tricksterstudio.endec.protocolVersionAlternatives
import me.maplesyrum.tricksterstudio.endec.recursiveStruct
import me.maplesyrum.tricksterstudio.endec.withAlternative
import me.maplesyrum.tricksterstudio.spell.executer.SpellInstruction
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder
import tree.maple.kendec.structEndecOf
import tree.maple.kendec.util.Optional

class SpellPart(var glyph: Fragment, var subParts: List<SpellPart>) : Fragment() {

    constructor(glyph: Fragment = PatternGlyph()) : this(glyph, ArrayList())

    override fun type(): FragmentType<out Fragment> {
        return FragmentType.SPELL_PART
    }

    override fun toString(): String {
        return "SpellPart"
    }

    companion object {
        val ENDEC: StructEndec<SpellPart> = recursiveStruct {
            StructEndecBuilder.of(
                Fragment.ENDEC.fieldOf("glyph") { obj: SpellPart -> obj.glyph },
                protocolVersionAlternatives(
                    mapOf(1 to it.listOf()),
                    withAlternative(
                        SpellInstruction.STACK_ENDEC.xmap(
                            { SpellUtils.decodeInstructions(it, listOf<Int>(), listOf<Fragment>(), Optional.empty<Fragment>()) },
                            (SpellUtils::flattenNode)
                        ), it
                    ).listOf()
                ).fieldOf ("sub_parts", SpellPart::subParts),
                ::SpellPart
            )
        }
    }
}
